File: 19.remove-nth-node-from-end-of-list.py
# @leet start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        fast = head
        slow = head
        i = 0
        while fast is not None:
            fast = fast.next
            if i > n:
                slow = slow.next
            else:
                i += 1
        if i == n:
            return head.next
        if slow.next is not None:
            slow.next = slow.next.next
        else:
            slow.next = None
        return head


# @leet end

File: 19.remove-nth-node-from-end-of-list.java
// @leet start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
		ListNode fast = head;
		ListNode slow = head;
		int i=0;
		while(fast!=null){
			fast=fast.next;
			if(i>n){
				slow=slow.next;
			}
			i++;
		}
		if(i==n)return head.next;
		slow.next=slow.next!=null?slow.next.next:null;
		return head;
    }
}
// @leet end
